### ISSUE TITLE
Calling to_der gives different result to original input der

### ISSUE THREAD

**Issue by adfoster-r7** [sdaubert/rasn1#24]
Test case 1 extracted from https://github.com/sdaubert/rasn1/issues/16

<details>

<summary>replication</summary>

```ruby
require 'rasn1'
require 'base64'

RASN1::Types.define_type('Uint32', from: RASN1::Types::Integer) do |value|
  (value >= 0) && (value < 2 ** 32)
end
RASN1::Types.define_type('Int32', from: RASN1::Types::Integer) do |value|
  (value >= -2 ** 31) && (value < 2 ** 31)
end
RASN1::Types.define_type('KerberosMicroseconds', from: RASN1::Types::Integer)
RASN1::Types.define_type('KerberosTime', from: RASN1::Types::GeneralizedTime)
RASN1::Types.define_type('KerberosFlags', from: RASN1::Types::BitString)
RASN1::Types.define_type('KerberosTicketFlags', from: RASN1::Types::KerberosFlags)

class KerberosString < RASN1::Types::IA5String
  # Important: Using GraphicString's id value of 27, instead of IA5String's ID of 22
  ID = 27

  # Get ASN.1 type
  # @return [String]
  def self.type
    'KerberosString'
  end
end

RASN1::Types.define_type('KerberosString', from: KerberosString)
RASN1::Types.define_type('KerberosRealm', from: RASN1::Types::KerberosString)

p RASN1::Model.singleton_methods

class PrincipalName < RASN1::Model
  sequence :PrincipalName,
           content: [
             int32(:name_type, explicit: 0, constructed: true),
             sequence_of(:name_string, RASN1::Types::KerberosString, explicit: 1, constructed: true)
           ]
end

class KrbError < RASN1::Model
  sequence :KrbError,
           class: :application,
           explicit: 30,
           content: [
             integer(:pvno, explicit: 0, constructed: true),
             integer(:msg_type, explicit: 1, constructed: true),
             kerberos_time(:ctime, explicit: 2, constructed: true, optional: true),
             kerberos_microseconds(:cusec, explicit: 3, constructed: true, optional: true),
             kerberos_time(:stime, explicit: 4, constructed: true),
             kerberos_microseconds(:susec, explicit: 5, constructed: true),
             uint32(:error_code, explicit: 6, constructed: true),
             kerberos_realm(:crealm, explicit: 7, constructed: true, optional: true),
             # vvv should not be output vvvv - asn1 spec: `cname           [8] PrincipalName OPTIONAL,`
             wrapper(model(:cname, PrincipalName), explicit: 8, constructed: true, optional: true),
             kerberos_realm(:realm, explicit: 9, constructed: true),
             wrapper(model(:sname, PrincipalName), explicit: 10, constructed: true),
             kerberos_string(:e_text, explicit: 11, constructed: true, optional: true),
             octet_string(:e_data, explicit: 12, constructed: true, optional: true),
           ]
end

valid_data =
  "\x7e\x81\x8d\x30\x81\x8a\xa0\x03\x02\x01\x05\xa1" \
  "\x03\x02\x01\x1e\xa4\x11\x18\x0f\x32\x30\x31\x34\x31\x32\x31\x34" \
  "\x32\x32\x34\x35\x32\x32\x5a\xa5\x05\x02\x03\x07\x5a\x47\xa6\x03" \
  "\x02\x01\x18\xa9\x0c\x1b\x0a\x44\x45\x4d\x4f\x2e\x4c\x4f\x43\x41" \
  "\x4c\xaa\x1f\x30\x1d\xa0\x03\x02\x01\x01\xa1\x16\x30\x14\x1b\x06" \
  "\x6b\x72\x62\x74\x67\x74\x1b\x0a\x44\x45\x4d\x4f\x2e\x4c\x4f\x43" \
  "\x41\x4c\xac\x30\x04\x2e\x30\x2c\x30\x16\xa1\x03\x02\x01\x0b\xa2" \
  "\x0f\x04\x0d\x30\x0b\x30\x09\xa0\x03\x02\x01\x17\xa1\x02\x04\x00" \
  "\x30\x12\xa1\x03\x02\x01\x13\xa2\x0b\x04\x09\x30\x07\x30\x05\xa0" \
  "\x03\x02\x01\x17".b

# expected = {
#   KrbError: {
#     pvno: 5,
#     msg_type: 30,
#     stime: DateTime.parse('2014-12-14T22:45:22+00:00'),
#     susec: 481863,
#     error_code: 24,
#     cname: {
#       name_type: nil, name_string: []
#     },
#     realm: 'DEMO.LOCAL',
#     sname: {
#       name_type: 1,
#       name_string: ['krbtgt', 'DEMO.LOCAL']
#     },
#     e_data: "0,0\x16\xA1\x03\x02\x01\v\xA2\x0F\x04\r0\v0\t\xA0\x03\x02\x01\x17\xA1\x02\x04\x000\x12\xA1\x03\x02\x01\x13\xA2\v\x04\t0\a0\x05\xA0\x03\x02\x01\x17".b
#   }
# }

krb_error = KrbError.parse(valid_data)

p krb_error
pp krb_error.to_h

$stderr.puts "input:"
$stderr.puts Base64.encode64(valid_data)
$stderr.puts "to_der:"
$stderr.puts Base64.encode64(krb_error.to_der)

if valid_data != krb_error.to_der
  raise 'failure - input der should match to_der result'
else
  puts 'success!'
end

```
</details>

Tests fail due to the asn1 output being different to the input:

```diff
Application 30 (1 elem)
-  SEQUENCE (8 elem)
+  SEQUENCE (9 elem) 
    [0] (1 elem)
      INTEGER 5
    [1] (1 elem)
      INTEGER 30
    [4] (1 elem)
      GeneralizedTime 2014-12-14 22:45:22 UTC
    [5] (1 elem)
      INTEGER 481863
    [6] (1 elem)
      INTEGER 24
+    [8] (1 elem)
+      SEQUENCE (2 elem)
+        [0] (1 elem)
+          INTEGER 0
+        [1] (1 elem)
+          SEQUENCE (0 elem)
    [9] (1 elem)
      GeneralString DEMO.LOCAL
    [10] (1 elem)
      SEQUENCE (2 elem)
        [0] (1 elem)
          INTEGER 1
        [1] (1 elem)
          SEQUENCE (2 elem)
            GeneralString krbtgt
            GeneralString DEMO.LOCAL
    [12] (1 elem)
      OCTET STRING (46 byte) 302C3016A10302010BA20F040D300B3009A003020117A10204003012A103020113A20B…
        SEQUENCE (2 elem)
          SEQUENCE (2 elem)
            [1] (1 elem)
              INTEGER 11
            [2] (1 elem)
              OCTET STRING (13 byte) 300B3009A003020117A1020400
                SEQUENCE (1 elem)
                  SEQUENCE (2 elem)
                    [0] (1 elem)
                      INTEGER 23
                    [1] (1 elem)
                      OCTET STRING (0 byte)
          SEQUENCE (2 elem)
            [1] (1 elem)
              INTEGER 19
            [2] (1 elem)
              OCTET STRING (9 byte) 30073005A003020117
                SEQUENCE (1 elem)
                  SEQUENCE (1 elem)
                    [0] (1 elem)
                      INTEGER 23
```

**Comment by adfoster-r7** [comment]
I've got a better workflow now for minifying the replication steps; Here's another reduced example for one of the other models that had the same issue:

<details>

<summary>replication</summary>

```ruby
require 'rasn1'
require 'base64'
require 'openssl'

RASN1::Types.define_type('Uint32', from: RASN1::Types::Integer) do |value|
  (value >= 0) && (value < 2 ** 32)
end
RASN1::Types.define_type('Int32', from: RASN1::Types::Integer) do |value|
  (value >= -2 ** 31) && (value < 2 ** 31)
end

p RASN1::Model.singleton_methods

class EncryptedData < RASN1::Model
  sequence :encrypted_data,
           content: [
             int32(:etype, explicit: 0, constructed: true),
             uint32(:kvno, explicit: 1, constructed: true, optional: true),
             octet_string(:cipher, explicit: 2, constructed: true)
           ]
end

class KdcReqBody < RASN1::Model
  sequence :KdcReqBody,
           content: [
             sequence_of(:etype, RASN1::Types::Int32, explicit: 8),
             wrapper(model(:enc_authorization_data, EncryptedData), explicit: 10, constructed: true, optional: true),
           ]
end

class KdcReq < RASN1::Model
  sequence :KdcReq,
           class: :application,
           explicit: 10,
           content: [
             wrapper(model(:req_body, KdcReqBody), explicit: 4, constructed: true)
           ]
end

valid_data = Base64.decode64(
  "ag0wC6QJMAeoBTADAgEX"
)

result = KdcReq.parse(valid_data)

p result
pp result.to_h

$stderr.puts "input:"
$stderr.puts Base64.encode64(valid_data)
$stderr.puts "to_der:"
$stderr.puts Base64.encode64(result.to_der)

if valid_data != result.to_der
  raise 'failure - input der should match to_der result'
else
  puts 'success!'
end
```
</details>

input_der (https://lapo.it/asn1js/#ag0wC6QJMAeoBTADAgEX) versus `result.to_der` (https://lapo.it/asn1js/#ahowGKQWMBSoBTADAgEXqgswCaADAgEAogIEAA) diff:

```diff
Application 10 (1 elem)
  SEQUENCE (1 elem)
    [4] (1 elem)
-      SEQUENCE (1 elem)
+      SEQUENCE (2 elem)
        [8] (1 elem)
          SEQUENCE (1 elem)
            INTEGER 23
+        [10] (1 elem)
+          SEQUENCE (2 elem)
+            [0] (1 elem)
+              INTEGER 0
+            [2] (1 elem)
+              OCTET STRING (0 byte)
```

**Comment by sdaubert** [comment]
Fix from #25 fixes your 2 examples above.

**Comment by adfoster-r7** [comment]
Thanks! Added another test case for a crash I ran into :+1:

https://github.com/sdaubert/rasn1/pull/26
